internal/fetchfixes/emb/argocd_manifest_builder.go

```diff
@@ -0,0 +1,223 @@
+      restartPolicy: OnFailure
+      serviceAccount: mcp-job-sa
+      serviceAccountName: mcp-job-sa
+      terminationGracePeriodSeconds: 30
+`
+
+const mcpPostHook = `---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  annotations:
+    argocd.argoproj.io/sync-wave: "1"
+    argocd.argoproj.io/hook: Sync
+    argocd.argoproj.io/hook-delete-policy: HookSucceeded
+  name: mcp-unpause-job
+  namespace: openshift-gitops
+spec:
+  template:
+    spec:
+      containers:
+        - image: registry.redhat.io/openshift4/ose-cli:v4.4
+          command:
+            - /bin/bash
+            - -c
+            - |
+              export HOME=/tmp/mcp
+              echo ""
+              echo -n "Un-pausing MachineConfigPools."
+              MCPS=({{ range .MCRoles }}"{{.}}" {{end}})
+              for MCP in "${MCPS[@]}"
+              do
+                oc patch machineconfigpools $MCP -p '{"spec":{"paused":false}}' --type=merge
+              done
+              echo "DONE"
+              echo -n "Waiting for the MachineConfigPools to converge."
+              sleep $SLEEP
+              MCPS=({{ range .MCRoles }}"{{.}}" {{end}})
+              for MCP in ${MCPS[@]}
+              do
+                until oc wait --for condition=updated --timeout=60s mcp $MCP
+                do
+                  echo -n "...still waiting for $MCP to converge"
+                  sleep $SLEEP
+                done
+              done
+              echo "DONE"
+          imagePullPolicy: Always
+          name: mcp-unpause-job
+          env:
+          - name: SLEEP
+            value: "5"
+      dnsPolicy: ClusterFirst
+      restartPolicy: OnFailure
+      serviceAccount: mcp-job-sa
+      serviceAccountName: mcp-job-sa
+      terminationGracePeriodSeconds: 30
+`
+
+type ArgoCDManifestBuilder struct {
+	needsMCManifests bool
+}
+
+func NewArgoCDManifestBuilder() ExtraManifestBuilder {
+	return &ArgoCDManifestBuilder{}
+}
+
+func getMCTemplates() map[string]*template.Template {
+	return map[string]*template.Template{
+		"sa-and-perms.yaml":  template.Must(template.New("sa-and-perms").Parse(saAndPerms)),
+		"mcp-pre-hook.yaml":  template.Must(template.New("pre-hook").Parse(mcpPreHook)),
+		"mcp-post-hook.yaml": template.Must(template.New("post-hook").Parse(mcpPostHook)),
+	}
+}
+
+func (amb *ArgoCDManifestBuilder) BuildObjectContext(fix, objOwner *unstructured.Unstructured) error {
+	if fix.GetKind() == "MachineConfig" {
+		amb.needsMCManifests = true
+	}
+
+	// TODO figure out dependency depth
+	if fixHasDependencies(fix, objOwner) {
+		amb.addWaveAnnotations(fix)
+	}
+	return nil
+}
+
+func (amb *ArgoCDManifestBuilder) addWaveAnnotations(obj *unstructured.Unstructured) {
+	if len(obj.GetAnnotations()) == 0 {
+		obj.SetAnnotations(make(map[string]string))
+	}
+	anns := obj.GetAnnotations()
+	// TODO: Replace with actual ArgoCD import?
+	// TODO: Keep track of dependencies and add higher waves if necessary
+	anns["argocd.argoproj.io/sync-wave"] = "2"
```

Comment:
Maybe an extra flag needed to specify the wave? (not sure what the sync-wave is).

Reply:
A sync wave gives the order in which operations happen. Given that this specific object has a dependency, we want to delay it until the dependency is met. Currently we only have one level of dependencies, so this is fine. But if in the future we create remediatoins that have multiple levels, this will fall short. Hence the TODO. The idea is that the `oc-compliance` script will be smart enough to do the math and provide a relevant wave number. At some point I'll come up with clever math to do this (and will probably do so by creating a dynamic graph of dependencies). But for now, 2 is as late as our dependencies go.

Reply:
That's cool. The rest of it looks good to me